[WPE][GTK][CMake] The path to bwrap and xdg-dbus-proxy should not be auto-detected when cross-compiling

clopez opened the pull request:
#### a84036c6d1d66d723f217a4c29eee76f2039a353
<pre>
[WPE][GTK][CMake] The path to bwrap and xdg-dbus-proxy should not be auto-detected when cross-compiling
<a href="https://bugs.webkit.org/show_bug.cgi?id=256679">https://bugs.webkit.org/show_bug.cgi?id=256679</a>

Reviewed by Adrian Perez de Castro.

When enabling -DENABLE_BUBBLEWRAP_SANDBOX=ON is needed to define to the build
the paths (full-paths) to the bwrap and xdg-dbus-proxy binaries.

The current CMake code is auto-detecting those paths by calling the CMake
function find_program(): so it is defining the paths to those programs with
the values from the host system.

But when cross-compiling that is wrong because the target binaries end with the
values for the paths from the host system which don&apos;t necessary have to match
the values from the target system.

I can&apos;t see how it will be possible to auto-detect the value that this programs
will have in the target system from the host system, so the only sane way of
dealing with this seems to be to give an error at configure time and ask for
those paths to be defined manually.

This patch changes the code to only try to auto-detect those binaries when no
cross-compiling.

Also update the default build parameters for the cross-building of targets
with cross-toolchain-helper to define the right paths that those targets will
have at run-time.

* Source/cmake/BubblewrapSandboxChecks.cmake:
* Tools/yocto/targets.conf:

Canonical link: <a href="https://commits.webkit.org/264244@main">https://commits.webkit.org/264244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2f14723cc49b8b58b7f23f1865aad656926da9c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8030 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6965 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6610 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9639 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6607 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8109 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5822 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13676 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5429 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8215 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6042 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6387 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5222 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6585 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5793 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1535 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9952 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6767 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/831 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6164 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1664 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/BubblewrapSandboxChecks.cmake:
```diff
@@ -1,42 +1,41 @@
-        endif ()
-        string(REGEX MATCH "[0-9]+\\.[0-9]+\\.[0-9]+" BWRAP_VERSION "${BWRAP_OUTPUT}")
-        if (NOT "${BWRAP_VERSION}" VERSION_GREATER_EQUAL "0.3.1")
-            message(FATAL_ERROR "bwrap must be >= 0.3.1 but ${BWRAP_VERSION} found")
```
This is removing the version check when not cross-compiling. I would rather leave it and move it to the appropriate location.

clopez replied:
Yes, that is on purpose. I don't think that version check makes sense anymore. Our oldest distro supported (Ubuntu-20.04) already ships brwap 0.4

Quoting from the changelog

> This patch also removes a version check for bwrap because that is not longer
> needed. The minimum version supported (which is 0.3.1) was released on Sep 2018
> and Ubuntu 20.04 already ships 0.4.0

aperezdc replied:
OTOH keeping the check is not only harmless, it may help people who accidentally try to build with older versions. In general we should have all the version checks possible done during the build configuration, if possible.

aperezdc commented on Source/cmake/BubblewrapSandboxChecks.cmake:
```diff
@@ -1,42 +1,41 @@
-        message(NOTICE
-            "***--------------------------------------------------------***\n"
-            "***  Cannot check Bubblewrap version when cross-compiling. ***\n"
-            "***  The target system MUST have version 0.3.1 or newer.   ***\n"
```
Also this notice gets remove without a replacement. It is important to let distributors know precisely which versions of the dependencies they must provide, so I would like to keep the note about the required minimum version when cross-compiling as well.

aperezdc commented on Source/cmake/BubblewrapSandboxChecks.cmake:
```diff
@@ -1,42 +1,41 @@
+    endif ()
+
+    # Do some extra sanity checks because the value of this paths has security implications if we get them wrong.
+    if (NOT "${BWRAP_EXECUTABLE}" MATCHES "^/.*/bwrap$")
```
There is an `IS_ABSOLUTE` conditional that can be used instead of regexes (faster, and clearer to understand):

```suggestion
    if (NOT IS_ABSOLUTE "${BWRAP_EXECUTABLE}")
```

I think it's not a good idea to check that the basename of the executable is `bwrap`. A distributor may install the program under a different name e.g. `bwrap-0.3` if they need different versions installed in parallel.

clopez replied:
It looks to me like a really hypothetical situation, but in that case they could workaround that by shipping the binaries in different directories like `/usr/local/bwrap-0.3/bin/bwrap` which is also the easiest way of doing it (the package build scripts allow one to easily install in a different path prefix but not to change the binary name, that would have to be done post-install manually)

I find more likely to happen that someone by mistake passes a wrong path like `/usr/bin/bwra` and then nothing works at run-time

aperezdc commented on Source/cmake/BubblewrapSandboxChecks.cmake:
```diff
@@ -1,42 +1,41 @@
+    if (NOT "${BWRAP_EXECUTABLE}" MATCHES "^/.*/bwrap$")
+        message(FATAL_ERROR "The value for BWRAP_EXECUTABLE is '${BWRAP_EXECUTABLE}'. It doesn't match the expected format (full path for a executable named 'bwrap'")
+    endif ()
+    if (NOT "${DBUS_PROXY_EXECUTABLE}" MATCHES "^/.*/xdg-dbus-proxy$")
```
Same here, better use `if (NOT IS_ABSOLUTE ...)`.

aperezdc reviewed (changes requested):
The idea is good overall, but I left a few comments, please take a look.

clopez committed:
[WPE][GTK][CMake] The path to bwrap and xdg-dbus-proxy should not be auto-detected when cross-compiling
https://bugs.webkit.org/show_bug.cgi?id=256679

Reviewed by Adrian Perez de Castro.

When enabling -DENABLE_BUBBLEWRAP_SANDBOX=ON is needed to define to the build
the paths (full-paths) to the bwrap and xdg-dbus-proxy binaries.

The current CMake code is auto-detecting those paths by calling the CMake
function find_program(): so it is defining the paths to those programs with
the values from the host system.

But when cross-compiling that is wrong because the target binaries end with the
values for the paths from the host system which don't necessary have to match
the values from the target system.

I can't see how it will be possible to auto-detect the value that this programs
will have in the target system from the host system, so the only sane way of
dealing with this seems to be to give an error at configure time and ask for
those paths to be defined manually.

This patch changes the code to only try to auto-detect those binaries when no
cross-compiling.

Also update the default build parameters for the cross-building of targets
with cross-toolchain-helper to define the right paths that those targets will
have at run-time.

* Source/cmake/BubblewrapSandboxChecks.cmake:
* Tools/yocto/targets.conf:

Canonical link: https://commits.webkit.org/264244@main
